File: 2022/scala/src/main/scala/AdventOfCode2022/Day12.scala
package AdventOfCode2022

import scala.annotation.tailrec

object Day12 {
  private val StartMarker: Char = 'S'
  private val EndMarker: Char = 'E'

  private case class Point(x: Int, y: Int) {
    def neighbours: Seq[Point] = Seq(Point(x + 1, y), Point(x - 1, y), Point(x, y + 1), Point(x, y - 1))
  }

  private def parse(input: Seq[String]): Map[Point, Char] =
    val points = for {
      y <- input.indices
      x <- input.head.indices
    } yield (Point(x, y), input(y)(x))

    points.toMap

  private def minDistance(to: Char, heights: Map[Point, Char]): Int = {
    def height(p: Point): Int = heights(p) match {
      case StartMarker => 0
      case EndMarker => 'z' - 'a'
      case h => h - 'a'
    }

    @tailrec
    def traverse(queue: List[Point], state: Map[Point, Int]): Map[Point, Int] = queue match
      case Nil => state
      case p :: tail =>
        val neighbours = p.neighbours
          .filter(heights.contains)
          .filter(!state.contains(_))
          .filter { n => height(p) - height(n) <= 1 }

        traverse(tail ++ neighbours, state ++ neighbours.map((_, state(p) + 1)).toMap)

    val start = heights.map(_.swap)(EndMarker)
    val distances = traverse(List(start), Map((start, 0)))
    distances.filter { x => heights(x._1) == to }.values.min
  }

  def partOne(input: Seq[String]): Int = minDistance(StartMarker, parse(input))

  def partTwo(input: Seq[String]): Int = minDistance('a', parse(input))

  def main(args: Array[String]): Unit = {
    val input = io.Source.fromResource("Day12.txt").getLines().toSeq
    println(s"Part one: ${partOne(input)}")
    println(s"Part two: ${partTwo(input)}")
  }
}

File: 2022/csharp/AdventOfCode2022/Day12.cs
namespace AdventOfCode2022;

internal class Day12
{
    private const char StartMarker = 'S';
    private const char EndMarker = 'E';
    private const char MinHeight = 'a';
    private const char MaxHeight = 'z';

    private readonly Dictionary<Point, char> _map;
    private readonly Dictionary<Point, int> _distances;
    private readonly Point _start;
    private readonly Point _end;

    public Day12(IEnumerable<string> input)
    {
        var heights = from line in input.Indexed()
                      from c in line.Value.Indexed()
                      select new { Point = new Point(c.Index, line.Index), Height = c.Value };

        _map = heights.ToDictionary(x => x.Point, x => x.Height);
        _start = heights.First(x => x.Height == StartMarker).Point;
        _end = heights.First(x => x.Height == EndMarker).Point;
        _distances = new Dictionary<Point, int> { [_end] = 0 };

        var queue = new Queue<Point>();
        queue.Enqueue(_end);

        while (queue.TryDequeue(out var point))
        {
            var neighbours = from neighbour in GetNeighbours(point)
                             where _map.ContainsKey(neighbour)
                             where !_distances.ContainsKey(neighbour)
                             where GetHeight(_map[point]) - GetHeight(_map[neighbour]) <= 1
                             select neighbour;

            foreach (var neighbour in neighbours)
            {
                _distances[neighbour] = _distances[point] + 1;
                queue.Enqueue(neighbour);
            }
        }
    }

    private static int GetHeight(char c) => c switch
    {
        StartMarker => 0,
        EndMarker => MaxHeight - MinHeight,
        _ => c - MinHeight,
    };

    public int PartOne()
    {
        return _distances[_start];
    }

    public int PartTwo()
    {
        return _distances.Where(x => _map[x.Key] == MinHeight).Min(x => x.Value);
    }

    private record struct Point(int X, int Y);

    private static IEnumerable<Point> GetNeighbours(Point point)
    {
        yield return new Point(point.X + 1, point.Y);
        yield return new Point(point.X - 1, point.Y);
        yield return new Point(point.X, point.Y + 1);
        yield return new Point(point.X, point.Y - 1);
    }
}

public class Day12Test
{
    private const string InputFile = "Day12.Input.txt";
    private const string ExampleFile = "Day12.Example.txt";

    [Theory]
    [FileData(ExampleFile, FileContents.StringPerLine, 31)]
    [FileData(InputFile, FileContents.StringPerLine, 330)]
    public void TestPartOne(IEnumerable<string> input, int expected)
    {
        var solution = new Day12(input);
        Assert.Equal(expected, solution.PartOne());
    }

    [Theory]
    [FileData(ExampleFile, FileContents.StringPerLine, 29)]
    [FileData(InputFile, FileContents.StringPerLine, 321)]
    public void TestPartTwo(IEnumerable<string> input, int expected)
    {
        var solution = new Day12(input);
        Assert.Equal(expected, solution.PartTwo());
    }
}
